api/types/networking.go

```diff
@@ -0,0 +1,230 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package types
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/gravitational/teleport/api/defaults"
+
+	"github.com/gravitational/trace"
+)
+
+// ClusterNetworkingConfig defines cluster networking configuration. This is
+// a configuration resource, never create more than one instance of it.
+type ClusterNetworkingConfig interface {
+	Resource
+
+	// GetClientIdleTimeout returns client idle timeout setting
+	GetClientIdleTimeout() time.Duration
+
+	// SetClientIdleTimeout sets client idle timeout setting
+	SetClientIdleTimeout(t time.Duration)
+
+	// GetKeepAliveInterval gets the keep-alive interval for server to client
+	// connections.
+	GetKeepAliveInterval() time.Duration
+
+	// SetKeepAliveInterval sets the keep-alive interval for server to client
+	// connections.
+	SetKeepAliveInterval(t time.Duration)
+
+	// GetKeepAliveCountMax gets the number of missed keep-alive messages before
+	// the server disconnects the client.
+	GetKeepAliveCountMax() int64
+
+	// SetKeepAliveCountMax sets the number of missed keep-alive messages before
+	// the server disconnects the client.
+	SetKeepAliveCountMax(c int64)
+
+	// GetSessionControlTimeout gets the session control timeout.
+	GetSessionControlTimeout() time.Duration
+
+	// SetSessionControlTimeout sets the session control timeout.
+	SetSessionControlTimeout(t time.Duration)
+
+	// CheckAndSetDefaults sets and default values and then
+	// verifies the constraints for ClusterNetworkingConfig.
+	CheckAndSetDefaults() error
+
+	// String returns string representation of cluster networking configuration.
+	String() string
+}
+
+// NewClusterNetworkingConfig is a convenience method to to create ClusterNetworkingConfigV2.
+func NewClusterNetworkingConfig(spec ClusterNetworkingConfigSpecV2) (ClusterNetworkingConfig, error) {
+	netConfig := ClusterNetworkingConfigV2{
+		Kind:    KindClusterNetworkingConfig,
+		Version: V2,
+		Metadata: Metadata{
+			Name:      MetaNameClusterNetworkingConfig,
+			Namespace: defaults.Namespace,
+		},
+		Spec: spec,
+	}
+
+	if err := netConfig.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &netConfig, nil
+}
+
+// DefaultClusterNetworkingConfig returns the default cluster networking config.
+func DefaultClusterNetworkingConfig() ClusterNetworkingConfig {
```

Comment:
is this the same as `NewClusterNetworkingConfig(ClusterNetworkingConfigSpecV2{})` ?
can this helper function be removed to keep default values in one place?

Reply:
Even if the resulting struct would be populated with the same data, the functions are not the same since `NewClusterNetworkingConfig` always returns an error. Initialization code that should work straightforwardly with no error handling expects `DefaultClusterNetworkingConfig` in this form.